### ISSUE TITLE
A "NullHandler" interface would be nice for custom Filters.

### ISSUE THREAD

**Issue by electrotype** [PebbleTemplates/pebble#223]
Currently, if I understand correctly, only `DefaultFilter` will be called if an expression evaluates to `null`.

```
if (filter instanceof DefaultFilter) {
    try {
        input = getLeftExpression().evaluate(self, context);
    } catch (AttributeNotFoundException ex) {
        input = null;
    }
} else {
    input = getLeftExpression().evaluate(self, context);
}
```

It would be nice if there was an interface our custom filters could implement so they are applied even if the result of the evaluation is `null`! My current workaround is to make my custom filter _extend_ `DefaultFilter`, but an interface would be way cleaner.

**Comment by mbosecke** [comment]
Can you provide more information about your custom filter? I suspect that this is too niche of a problem that any solution is probably not worth the overhead.

**Comment by electrotype** [comment]
In the framework I work on, I use Pebble to output informations if fields from a submitted Form contain errors. 

For example, If "_myForm.user.company.title_" is invalid, I want to ouput an error message. But it is possible that "myForm.user.company" doesn't exist yet... In that case, I simply don't want to output anything. Note that I still want to use `strictVariables` everywhere in the application.

Using my workaround, I do something like :
`
{{myForm.user.company.title | fieldErrors}}
`

Since the `fieldErrors` Filter extends `DefaultFilter`, the Java code is called even if "_myForm.user.company_" doesn't exist yet, and I then simply output an empty String. If this field exist, I can validate it an output an error if required.

If I try to do something similar but using a function, for example :

`{{fieldErrors(myForm.user.company.title)}}` 

then an exception is thrown when "_myForm.user.company_" doesn't exist.

In other words, I need the `myForm.user.company.title` expression to be evaluated, and my Java code to always be called even if one property doesn't exist. Is there another way to achieve that than my workaround?

I can submit a Pull Request for that, it's trivial to implement.

**Comment by mbosecke** [comment]
So, essentially you want to continue using `strictVariables` but in this particular instance you do not want the `AttributeNotFoundException` to be thrown. 

I'm reluctant to allow this exception to be suppressed by arbitrary filters because it defeats the purpose of using `strictVariables` in the first place and it makes it harder to read a template when it's not obvious which filters suppress the exception and which ones don't. In fact, I don't even like that the `DefaultFilter`  suppresses this exception. 

My preference is to maintain status quo for now, but I'll keep this task open so that it can be revisited for the next major version of Pebble when we don't mind breaking backwards compatibility. At that time, I would want the solution to be *crystal clear* when reading the template that `strictVariables` has been temporarily turned off. Perhaps a dedicated filter for that reason: `{{ user.company.title | strictVariables(false) | fieldErrors }}`.

Some possible workarounds that you can implement with the current system:
- Extend `DefaultFilter` like you mentioned
- Use the `DefaultFilter` for the sake of suppressing the exception prior to feeding the input into your own filter. I.e.: `{{ myForm.user.company.title | default(null) | fieldErrors }}`. Looks awkward and not obvious what the intent is.
- Use ternary expressions: `{{ (myForm.user.company == null? null : myForm.user.company.title) | fieldErrors }}`. This is probably the "cleanest" solution, although ternary expressions are verbose.

**Comment by electrotype** [comment]
Thanks for that reply. I understand your concerns, what I don't see how it would break any backwards compatibility or be confusing. You simply do:

`
if (filter instanceof AttributeNotFoundExceptionHandler) {
`

Instead of 
`
if (filter instanceof DefaultFilter) {
`

and you make `DefaultFilter` the only one to implement that interface. Then developers have the freedom to implement the interface for there custom filters, if it's required.

But at least there are a couple of workarounds, so it ok for me! Thanks for your help.

**Comment by mbosecke** [comment]
The "confusing" problem with your proposed solution is when you read `{{ user.company.title | myCustomFilter }}` and you know that `strictVariables` is true, you have no idea whether or not it's null-safe until you dig open the source code for `myCustomFilter`. This is too big of a problem to go forward with that solution and I suspect that any other solution will require breaking backwards compatibility with the `DefaultFilter`.

**Comment by electrotype** [comment]
I see... At least, don't make `DefaultFilter` final please! ;-)

**Comment by nikku** [comment]
There is another option to consider: Why not add a syntax that allows users to safely query for null values, even if `strictVariables` is true?

____Example____

Let's assume the `user` attribute is either set or not. 

The expression `{% if user != null %}...{% endif %}` will fail in strict mode.

If I am absolutely aware of the fact that the user may be optional, I could make that explicit, i.e. via `{% if user? != null %}...{% endif %}` to relax the `strictVariables` setting.

The filter mentioned above would simply reduce to the following then:

```
{{ myForm.user.company.title? | fieldErrors}} 
```

**Comment by electrotype** [comment]
I still think that libraries and frameworks should ideally have the possibility to create functions and filters that are able to manage non existing attributes. 

I currently have my hands _deep_ into form validation, for the framework I work on, and I must say that without that "extending the `DefaultFilter`" hack, Pebble would probably be too limiting for what I need.

Let me give you more details about that already discussed "validation" use-case.

Displaying an HTML form would look like this in my framework (simplified example) :

    <form method="post">
        <input type="text" 
               class="{{validation.myForm.user.company.title | validationClass}}" 
               name="myForm.user.company.title" 
               value="{{myForm.user.company.title | fieldValue}}">
        {{validation.myForm.user.company.title | validationMessages}}
        <button type="submit">Submit</button>
    </form>

- I'm not fan of frameworks that _hide_ the HTML generation by providing some "magic" tag/function which output the entire Form by itself! That's why I give the possibility to manually add :
    - The value of the field itself (using the `fieldValue` filter)
    - The css class for the input, depending of the validation result (using the `validationClass` filter)
    - The validation messages associated to a particular input (using the `validationMessages` filter)

- Plain Maps are used to access both the fields values _and_ the validation results. This makes it very easy for a developer to use any templating engine (they all handle plain Maps) :
    - The value of the company title is taken from a plain Map : `myForm.user.company.title`.
    - The potential validation messages are also inside a plain Map, but a different one : `validation.myForm.user.company.title`. 

I don't want to say to the developers : "You have to make sure that there is no validation messages associated to a field before trying to output them.", that would be very ugly in my opinion... The `{{validation.myForm.user.company.title | validationMessages}}` way (using the "extending the `DefaultFilter`" hack) is not bad in that regard : it is straightforward and the developer doesn't have to worry about the fact that there may not be any validation messages for some fields.

Frameworks such as Spring MVC (with JSP for example) provide complex `tags` which handle such validation output, [form:errors](http://docs.spring.io/spring/docs/current/spring-framework-reference/html/view.html#view-jsp-formtaglib-errorstag) for example. The developers don't need to check if there _are_ errors or not to use this tag...

In other words, I still think that preventing third-parties to provide functions and filters that handle non existing attributes is somewhat limiting.

_________
That said, I see another option though : If the "expression parsing" algorithm Pebble uses was available using an API, then everything could work too. For example, if there was this method available :

`
EvaluationContext#evaluate(String expression)
`

Then we could write a function (or a filter) which would looks like : `{{validationMessages("validation.myForm.user.company.title")}}` (notice the quotes!)

And the function/filter could evaluate this expression _by itself_, using the `evaluate(...)` method AND it could handle non existing values, if required.

Sorry for the wall of text! :-)

**Comment by electrotype** [comment]
I now have a [live example](https://www.spincast.org/documentation#forms_validation_filters) of what I'm talking about. Have a look at the first filter, `validationMessages()`, for example.

I would have prefered to use _functions_ there, but since there is no way to make them "attribute does not exist" tolerant, I used filters with the "extends `DefaultFilter`" hack [[code](https://github.com/spincast/spincast-framework/blob/master/spincast-plugins/spincast-plugins-pebble-parent/spincast-plugins-pebble/src/main/java/org/spincast/plugins/pebble/SpincastPebbleExtensionDefault.java#L317)].

I want the users to be able to have `strictVariables` _on_ in their application. But the first time the form is displayed, the `validation` root variable doesn't exist, and throwing an exception here is not acceptable (you can see [here](https://www.spincast.org/documentation#forms_validation_validating) why this `validation` variable doesn't exist the fist time the form is displayed).

Anyway, it's just to show why a `AttributeNotFoundExceptionHandler` interface (or something similar) would be beneficial in some cases, imo.

**Comment by ebussieres** [comment]
Related to #245
